### DISCUSSION TITLE
Can we add constraints on relationships between parameters for TPE sampler?

### DISCUSSION THREAD

**Discussion by vangarp** [optuna/optuna#3965] [Q&A] [answered]
Hi folks, I saw that with release 3.0.0, TPE sampler can accept a constrints function where we can tell if a particular trial is infeasible. However it seems like this is only evaluated after the trial is complete. We are using TPE sampler in a study in ask-tell mode and was wondering if there is a way to have the ask() method return a trial object such that it will only suggest assignments which satisfy the constraints?
Ex. I have a study with parameters x and y both are int. I have a requirement that x must always be greater than y and both are within some range say 100-200. How can I make sure that I only ever run trials where the assignments of x and y satisfy the constraint of x > y? As per my understanding currently if I use constrints_func in TPE sampler, it will still execute trials when x can be <= y and then as we report it's score that trial will be discarded while reporting best_trials.

**Comment by nzw0301** [comment, marked as answer]
> there is a way to have the ask() method return a trial object such that it will only suggest assignments which satisfy the constraints?

Unfortunately, the answer is no because Optuna supports only soft constraints as mentioned in https://optuna.readthedocs.io/en/stable/faq.html#how-can-i-optimize-a-model-with-some-constraints.
